### PR TITLE
Fix pillar propagation in salt-ssh when overriding pillar in module.run + state.apply

### PR DESCRIPTION
**Pull request by mateiw**
### What does this PR do?

Add option to merge pillar with `opts['pillar']` during pillar compilation. The options is `False` by default and only set by `salt-ssh` in order to allow for the pillar to be propagated when applying a state using `module.run` + `state.apply` and overriding the pillar explicitly. 

This problem doesn't happen with regular, non-ssh minions.

### What issues does this PR fix or reference?

#47501

### Previous Behavior
If `pillar` is specified in `module.run` + `state.apply` it replaces any previous pillar data completely in the state that is applied instead of merging the value with the existing data.
E.g.
```yaml
call_show:
  module.run:
    - name: state.apply
    - mods:
      - showpillar
    - kwargs: 
          pillar:    # <= replaces existing pillar data completely in the showpillar state
              test: "foo bar"
```

### New Behavior
`salt-ssh` sets the option `pillar_compile_merge_global` to `True` in order to for the pillar compilation to merge the new data with the existing one.

### Tests written?

No

### Commits signed with GPG?

No

### CODE REVIEW

**Review by gtmanfred — APPROVED** [review]

File: salt/client/ssh/__init__.py [hunk 1]
```diff
@@ -911,6 +911,7 @@ def __init__(
                     'id': self.id,
                     'sock_dir': '/',
                     'fileserver_list_cache_time': 3,
+                    'pillar_compile_merge_global': True
```
Comment (isbm) [inline salt/client/ssh/__init__.py]: This gets set to `True` all the time and ignores _any kind_ of settings. I would be OK to set it to `True` by default, but completely ignoring the opposite value of it isn't the great idea. And maybe shorter name? like `merge_pillar`.

**Review by isbm — CHANGES_REQUESTED** [review]

**Comment by mateiw** [comment]
Thanks for the suggestion @isbm. Renamed to `pillar_merge_global` and made it possible to override the default value using `minion_opts` in the roster file or globally using `ssh_minion_opts` in the master config.

File: salt/client/ssh/__init__.py [hunk 2]
```diff
@@ -901,6 +901,7 @@ def __init__(
         self.minion_opts = {
                     'grains_cache': True,
                     'log_file': 'salt-call.log',
+                    'pillar_merge_global': True,
```
Comment (isbm) [inline salt/client/ssh/__init__.py]: I'd remove this and put it do the default `ssh_.....` options that comes from `ssh_minion_opts`. Besides, I would use conventional name `ssh_<do>_<what>`. That would be `ssh_merge_pillar`, and set it to `True` by default. Anyone who wants it void can set it to `False` for Salt SSH.

**Review by isbm — CHANGES_REQUESTED** [review]
@mateiw some more updates.

**Review by isbm — APPROVED** [review]
@mateiw thanks, perfect! Now we're good to go.

**Review by terminalmage — CHANGES_REQUESTED** [review]
This looks good, but we should document the new config option in doc/ref/configuration/minion.rst (with a `.. versionadded:: 2018.3.2` to note when the option was added).

**Comment by mateiw** [comment]
@terminalmage I added the documentation for this new option.

File: doc/ref/configuration/minion.rst [hunk 3]
```diff
@@ -3219,3 +3219,31 @@ URL of the repository:
 Replace ``<commit_id>`` with the SHA1 hash of a commit ID. Specifying a commit
 ID is useful in that it allows one to revert back to a previous version in the
 event that an error is introduced in the latest revision of the repo.
+
+``ssh_merge_pillar``
+----------------------
```
Comment (terminalmage) [inline doc/ref/configuration/minion.rst]: Can you delete the last two dashes here? Sphinx will warn on the docs build when the dashed line is longer than the line above it.

**Review by terminalmage — CHANGES_REQUESTED** [review]
Thanks for adding the docs! Just one minor tweak needed here.

**Review by terminalmage — APPROVED** [review]

**Pull request merged**